### ISSUE TITLE
Question: why is the Visitor trait limited to statements, relations & expressions?

### ISSUE THREAD

**Issue by freshtonic** [sqlparser-rs/sqlparser-rs#934]
What is the reason for that particular design decision versus providing a more general `Visitor` implementation?

Such as:

1. exposing a trait method (pre + post) for every AST node type, or
2. exposing a just two methods (pre + post) with a signature like `fn pre_visit(&mut self, node: &AstNode) -> ControlFlow<Self::Break>` - where `AstNode` is an enum with a wrapper variant for every AST node type found in `src/ast/mode.rs` that can be `match`ed against.

Would the maintainers be interested in a PR that implements one of the above two approaches? I have some thoughts about how such a PR could be implemented without breaking backwards compatibility.

**Comment by alamb** [comment]
> What is the reason for that particular design decision versus providing a more general Visitor implementation?

The core reasons are  to keep this crate easy to extend with new sql syntax by casual contributors and limited maintenance bandwidth. 

The current visitor pattern / macros require a single new `#[derive]` call 

In terms of your proposed implementation, it seems reasonable. I would be interested in knowing how it would work when a contributor added a new node, like https://github.com/sqlparser-rs/sqlparser-rs/pull/897

How would we ensure that these new nodes were added to the `AstNode` enum? What would happen if we forgot to do so?

perhaps @lovasoa  and @tustvold  have some other perspectives

Basically I think a general purpose rewriter / visitor would be a great addition, as long as we don't make it too hard to add new code to this repo

**Comment by alamb** [comment]
> sqlparser-rs is at the very core of [SQLPage](https://github.com/lovasoa/SQLpage), so I can reiterate my proposition: I wouldn't mind being added as a maintainer here :)

Indeed -- I am looking for help maintaining it for sure. I think you can do all the time consuming parts of maintenance (PR reviews, answering questions / tickets, etc) without having write access to the repository.

I am waiting for someone to actually show the initiative to do this work (there are plenty of people willing to write code). 

I had a promising candidate in @AugustoFKL  (see https://github.com/sqlparser-rs/sqlparser-rs/issues/808) but sadly there were some circumstances that required them to stop working here

I will write up a discussion explaining the current state of affairs

**Comment by freshtonic** [comment]
> In terms of your proposed implementation, it seems reasonable. I would be interested in knowing how it would work when a contributor added a new node, like https://github.com/sqlparser-rs/sqlparser-rs/pull/897
>
> How would we ensure that these new nodes were added to the AstNode enum? What would happen if we forgot to do so?

That's a great question. Assuming the parent node of the new node has a derived Visitor instance then compilation would fail  due to the derived code not being able to find a matching `AstNode` variant, which is not ideal.

Ideally, the `AstNode` enum would be generated automatically by walking the types starting with `sqlparser::ast::Statement` but that would require some deep macro voodoo.

**Comment by lovasoa** [comment]
You don't need to create a huge enum. You can use the built in Any trait. Have a look at my PR ;)

**Comment by freshtonic** [comment]
@lovasoa depends on if you want to `match` on it. I'll take a look at you PR though.

**Comment by alamb** [comment]
FWIW https://github.com/sqlparser-rs/sqlparser-rs/pull/951 added visiting TableFactor as well

**Comment by freshtonic** [comment]
Update: I've been experimenting with various implementation strategies to generalise the AST visitor. Nothing to ready to show yet, but this is business critical for my company so it's a high priority to figure out something that works well.

I'll update this issue with a link to a draft PR hopefully in a matter of days.

**Comment by freshtonic** [comment]
Progress update - I have a working generalised visitor implementation. I'll put up a PR later this week when I've done some more testing.

**Comment by osaton** [comment]
Is this still happening? Generalised visitor implementation would be appreciated.

**Comment by freshtonic** [comment]
@osaton @alamb yes this is still happening. Apologies for the lack of comms.

TL;DR I ended up taking quite a different approach to what I originally planned, due to our (the company I work for) evolving needs.

The comprehensive Visitor implementation will be initially released as its own crate.

Summary of features so far:

- A `Visitor` implementation is notified of every type in the `sqlparser` AST (200+ types when you take into account the `Vec`, `Box` and `Option` combinations, plus primitives).
- The `Visitor` trait itself is generic over the AST node type of interest (there aren't 200+ trait methods - in fact there are just two: `enter` and `exit`).
- The AST traversal is performed in an optimal order for semantic analysis workloads (instead of the order of AST fields in the source). Specifically, AST nodes that can refer to other AST nodes are visited _after_ the AST node that introduces the referent - e.g. a `FROM` clause is visited before the a projection that refers to columns in the tables introduced by the `FROM` clause. This cuts down on the number of passes required for semantic analysis workloads.

In order to pull that off, I needed to write a `build.rs` script to analyse `sqlparser`'s sources because hand writing AST traversal for all of the AST types would have been a _lot_ of work.

I need to write up some docs but I would love some feedback on the code before I publicly publish the repo.

Would you both be keen on taking a look later this week?

**Comment by freshtonic** [comment]
Three features that I _really_ wanted from `sqlparser` while writing the enhanced Visitor code:

1. Optional feature to have a unique ID baked into all AST nodes

It's possible for multiple AST nodes at different places in the tree that have the same type to have the same hash and compare as equal. This seems counter intuitive, but two syntactically identical nodes can be different _semantically_.

Not having robust (and cheap) means of uniquely identifying a node makes building HashMaps of derived metadata keyed by node particularly tricky.  

2. Optional feature to have span information baked into all AST nodes

This would be super useful for debugging and reporting errors nicely (like `miette` does).

3. Type per enum variant (no fields) everywhere

It would be nice to not have to match an enum in order to pluck out a variant. Instead I'd like be able to implement `Visit<InsertStatement>`.

1 & 2 could be solved by wrapping every struct field or enum variant field in a `Node<T>` type that `Deref`s to `T` but also contains a unique ID and span info. Because of the `Deref` it _might_ be possible to do this without breaking backwards compatibility but I haven't actually tried it.

**Comment by osaton** [comment]
@freshtonic Your visitor crate sounds perfect for my use case. I would love to try it but I'm not sure if I'm able to give any constructive criticism about the code as I'm fairly new to Rust and systems programming languages in general.

> 1. Optional feature to have a unique ID baked into all AST nodes
> 
> It's possible for multiple AST nodes at different places in the tree that have the same type to have the same hash and compare as equal. This seems counter intuitive, but two syntactically identical nodes can be different _semantically_.
> 
> Not having robust (and cheap) means of uniquely identifying a node makes building HashMaps of derived metadata keyed by node particularly tricky.
> 
> 2. Optional feature to have span information baked into all AST nodes
> 
> This would be super useful for debugging and reporting errors nicely (like `miette` does).
> 
> 3. Type per enum variant (for `Statement` and `Expr` at least, maybe some additional other types)

I don't know if this can be done without breaking changes, but having both 1 and 2 would be amazing.

**Comment by freshtonic** [comment]
@ryb73 I built `sqltk` on top of `sqlparser` which provides an alternate take on the `Visitor` trait. You can find it here [`sqltk`](https://crates.io/crates/sqltk).

Note that the docs building currently fails on docs.rs due to very clunky build step to work around being unable to derive traits for foreign types in Rust.

The workaround I'm considering is to fork `sqlparser` as `sqltk-parser` in the workspace (which gratuitous credit given) and maintain compatibility with upstream `sqlparser`.

We're using it in production at CipherStash so it's "production ready" in that sense but we'd appreciate some community feedback and PRs to fix/improve it :)

**Comment by iffyio** [comment]
Yeah I think option 2 above sounds reasonable!

**Comment by ramnes** [comment]
FWIW I quickly implemented option 1 [here](https://github.com/formalco/datafusion-sqlparser-rs/commit/3ab0faf6247287c9c31abd048ad8ed93f105fe01) in the meantime, if you want to take a look. That's indeed a large chunk of boilerplate code.

Anyone wants to work on option 2? Otherwise I may be able to land something next week, but I don't want to guarantee anything.

**Comment by freshtonic** [comment]
@ramnes anything I can do to help you make [your Visitor changes](https://github.com/formalco/datafusion-sqlparser-rs/commit/3ab0faf6247287c9c31abd048ad8ed93f105fe01) into a PR?

@alamb is there a realistic change of @ramnes work being merged?

If it lands, it potentially means `sqltk` no longer needs to exist. The only other unique feature of `sqltk` is a traversal order guarantee, which it turns out I don't need.

**Comment by ramnes** [comment]
> Two options for a generalised Visitor trait come to mind:
> 
> 1. expose pre + post trait method variants for every AST node type, or
> 2. expose only two trait methods (pre_visit + post_visit) with signatures like fn pre_visit(&mut self, node: &AstNode) -> ControlFlow<Self::Break> - where AstNode is an enum with a wrapper variant for every AST node type found in src/ast/mod.rs and can be matched against.

I meant option 2 here. Isn't it what @alamb and @iffyio meant as well? 😅

**Comment by freshtonic** [comment]
@ramnes thank you for clarifying!

I'm 100% in favour of any style of `Visitor` implementation that covers all AST node types in `sqlparser`. I'm less fussy than I used to be about the particular approach taken. These are the "obvious" approaches to me:

1. Big `Visitor` trait (enter/exit or pre/post methods for _every_ AST node type)
2. Small `Visitor` trait (just enter/exit methods, e.g. `fn enter<N: Visitable>(&mut self, node: &N) -> ControlFlow`
3. Small generic `Visitor<N: Visitable>` trait (just enter/exit methods, e.g. `fn enter(&mut self, node: &N) -> ControlFlow` 

Option 1 is (subjectively) ugly but gets the job done.

Option 2 has a cleaner trait but implementations must downcast the node to do anything useful with it.

Option 3 is just as clean as 2 and downcasting is no longer required in every implementation, _but_ `Visitor` dispatch no longer works because the AST traversal has to be able to dispatch _any_ node type to a `Visitor`, so you still need an additional dispatch mechanism. This is where the enum based dispatch could help.

The approach to `Visitor` I ended up going with in `sqltk` is essentially Option 2 above and looks like [this](https://docs.rs/sqltk/latest/sqltk/trait.Visitor.html).

I have been thinking about alternative approach to avoid all of the explicit casting and just generate the code.

Here's the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=f1d842d0b5a9d4af39a4684ffc41265a)

**Comment by freshtonic** [comment]
@ramnes oh and one more thing that's very important.

For workloads such as analysing an AST (in my case, deriving type information for a subset of node types) a `Visitor` implementation has to be able to associate information with AST nodes, so an implementation would have, say, a field like `derived_node_info: HashMap<&Expr, SomeDerivedInfo>`.

The problem is that `&Expr` _must_ have a lifetime and that lifetime must be at least as long as the parsed AST lasts.

That means the `Visitor` trait has to have some lifetime associated with the trait itself (e.g. `Visitor<'ast>`) and the enter/exit methods must use that lifetime for the node references, e.g. `fn enter(&mut self, node: &'ast N) ...`.
